rusk/src/lib/ws.rs

```diff
@@ -76,122 +101,316 @@ async fn listening_loop<N: Network, DB: database::DB>(
+                        let _ = stream.close(Some(CloseFrame {
+                            code: CloseCode::Error,
+                            reason: Cow::from("Failed sending response"),
+                        })).await;
+                        break;
+                    }
+                }
+
+                msg = stream.next() => {
+                    match msg {
+                        Some(msg) => match msg {
+                            Ok(msg) => match msg {
+                                // We received a request and should spawn a new tas to handle it.
+                                Message::Text(msg) => {
+                                    match serde_json::from_str(&msg) {
+                                        Ok(req) => {
+                                            task::spawn(handle_request(
+                                                sources.clone(),
+                                                req,
+                                                responder.clone(),
+                                            ));
+                                        },
+                                        Err(err) => {
+                                            let _ = stream.close(Some(CloseFrame {
+                                                code: CloseCode::Error,
+                                                reason: Cow::from(format!("Failed deserializing request: {err}")),
+                                            })).await;
+                                            break;
+                                        }
+                                    }
+                                }
+                                // Any other type of message is unsupported.
+                                _ => {
+                                    let _ = stream.close(Some(CloseFrame {
+                                        code: CloseCode::Unsupported,
+                                        reason: Cow::from("Only text messages are supported"),
+                                    })).await;
+                                    break;
+                                },
+                            }
+                            // Errored while receiving the message, we will
+                            // close the stream and return a close frame.
+                            Err(err) => {
+                                let _ = stream.close(Some(CloseFrame {
+                                    code: CloseCode::Error,
+                                    reason: Cow::from(format!("Failed receiving message: {err}")),
+                                })).await;
+                                break;
+                            }
+                        }
+                        // The stream has stopped producing messages, and we
+                        // should close it and stop. The client likely has done
+                        // this on purpose, and it's a part of the normal
```

Comment:
Looking at a [`futures::Stream`](https://docs.rs/futures/latest/futures/stream/trait.Stream.html) we can see that the specification says that it will return `None` when the stream is exhausted. What "exhausted" means in the context of `WebSocketStream<T>` is really up to [the implementation](https://docs.rs/tokio-tungstenite/0.19.0/src/tokio_tungstenite/lib.rs.html#276-307), but looking at it we can see that `None` is returned in the cases that a connection has been closed or is otherwise unusable.